internal/dispatcher/dispatcher-operator.go

```diff
@@ -0,0 +1,440 @@
+				d.ModifiedHandler(obj, gvr)
+			case watch.Deleted:
+				klog.Infof("the resource %s %s has been deleted from local cluster ", obj.GetName(), gvr.String())
+				d.DeletedHandler(obj, gvr)
+			}
+		}
+	}
+}
+
+func (d *DispatcherReconciler) StartWatchers() {
+	//for each resource check if there is already a running watcher
+	for _, r := range d.RegisteredResources {
+		//if there is not then start one
+		if _, ok := d.RunningWatchers[r.String()]; !ok {
+			stop := make(chan bool)
+			d.RunningWatchers[r.String()] = stop
+			go d.Watcher(r, stop)
+			klog.Infof("starting watcher for resource: %s", r.String())
+		} else {
+			//here we check if the channel is closed and if so then we start again the watcher
+			if !isOpen(d.RunningWatchers[r.String()]) {
+				//it means that the channel is closed we need to restart the watcher
+				stop := make(chan bool)
+				d.RunningWatchers[r.String()] = stop
+				go d.Watcher(r, stop)
+				klog.Infof("starting watcher for resource: %s", r.String())
+			}
+		}
+	}
+}
+
+//Stops all the watchers for the resources that have been unregistered
+func (d *DispatcherReconciler) StopWatchers() {
+	for _, r := range d.UnregisteredResources {
+		if ch, ok := d.RunningWatchers[r]; ok {
+			if !isOpen(ch) {
+				//it means that the channel is closed we need only to delete the entry in the map
+				delete(d.RunningWatchers, r)
+			} else {
+				close(ch)
+				delete(d.RunningWatchers, r)
+				klog.Infof("stopping watcher for resource: %s", r)
+			}
+		}
+	}
+}
+
+func (d *DispatcherReconciler) CreateResource(client dynamic.Interface, gvr schema.GroupVersionResource, obj *unstructured.Unstructured, clusterID string) error {
+	//check if the resource exists
+	name := obj.GetName()
+	namespace := obj.GetNamespace()
+	klog.Infof("creating resource %s of type %s on cluster %s", name, gvr.String(), clusterID)
+	r, found, err := d.GetResource(client, gvr, name, namespace, clusterID)
+	if err != nil {
+		klog.Errorf("an error occurred while getting resource %s of type %s on cluster %s: %s", name, gvr.String(), clusterID, err)
+		return err
+	}
+	if found {
+		//the resource already exists check if the resources are the same
+		if areEqual(obj, r) {
+			klog.Infof("resource %s of type %s already exists on cluster %s ", obj.GetName(), gvr.String(), clusterID)
+			return nil
+		} else {
+			//if not equal delete the remote one
+			err := client.Resource(gvr).Namespace(namespace).Delete(context.TODO(), name, metav1.DeleteOptions{})
+			if err != nil {
+				klog.Errorf("an error occurred while deleting resource %s on cluster %s: %s", name, clusterID, err)
+				return err
+			}
+		}
+	}
+	//if we come here it means that we have to create the resource on the remote cluster
+	spec, b, err := unstructured.NestedMap(obj.Object, "spec")
+	if !b || err != nil {
+		klog.Errorf("an error occurred while processing the 'spec' of the resource %s: %v ", name, err)
+		return err
+	}
+	remRes := &unstructured.Unstructured{
```

Comment:
Do we need some annotations to say that the resource have been reflected? From the following resource, it is not clear to me how to detect which resources have been reflected and which are "native"...